src/main/java/com/amazon/ion/impl/bin/WriteBuffer.java

```diff
@@ -100,10 +100,11 @@ public void truncate(final long position)
     {
         final int index = index(position);
         final int offset = offset(position);
-        final Block block = blocks.get(index);
-        this.index = index;
-        block.limit = offset;
-        current = block;
+        while (this.index != index) {
+            blocks.remove(this.index--);
+        }
```

Comment:
🗺️ This fixes a bug in `WriteBuffer` that could potentially lead to corrupted data. Basically, if we change the `index` without either removing or resetting the blocks higher than `index`, then when we start writing again, and we get to one of those blocks, we wouldn't have started at position 0. For example:
```
| Block 0 | Block 1 |
|=========|=========|
|         |         |  (1) write "ABCDEF" to the buffer
| A B C D | E F _ _ |
|         |         |  (2) truncate to position 3
| A B C _ | E F _ _ |  <-- Notice that while block 0 is reset to the correct position, there 
|         |         |      was no clean up done in block 1
|         |         |      
|         |         |  (3) write "123" to the buffer
| A B C 1 | E F 2 3 |  !!! Block 1 resumed at offset 2, so we still 
|         |         |      have the "EF" that we tried to truncate earlier.
```

Reply:
How long has this bug existed? Does it require patching in master and/or previous versions?

Reply:
I think it's been there all along, but I haven't looked into the potential impact. Since it's been there all along, it's even possible that the current implementation has a way to avoid the buggy behavior by cleaning the blocks at some other time.